python/lsst/pipe/base/mermaid_tools.py

```diff
@@ -0,0 +1,431 @@
+# the Free Software Foundation, either version 3 of the License, or
+# (at your option) any later version.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+"""Module defining few methods to generate Mermaid charts from pipelines or
+quantum graphs.
+"""
+
+from __future__ import annotations
+
+__all__ = ["graph2mermaid", "pipeline2mermaid"]
+
+import html
+import re
+from collections.abc import Iterable
+from typing import TYPE_CHECKING, Any, Iterable
+
+from lsst.daf.butler import DatasetType, DimensionUniverse
+
+from . import connectionTypes
+from .connections import iterConnections
+from .pipeline import Pipeline
+
+if TYPE_CHECKING:
+    from lsst.daf.butler import DatasetRef
+    from lsst.pipe.base import QuantumGraph, TaskDef
+
+
+def _datasetRefId(dsRef: DatasetRef) -> str:
+    """Make a unique identifier string for a dataset ref based on its name and
+    dataId."""
+    dsIdParts = [dsRef.datasetType.name]
+    dsIdParts.extend(f"{key}_{dsRef.dataId[key]}" for key in sorted(dsRef.dataId.required.keys()))
+    return "_".join(dsIdParts)
+
+
+def _makeDatasetNode(dsRef: DatasetRef, allDatasetRefs: dict[str, str], file: Any) -> str:
+    """Create a Mermaid node for a dataset if it doesn't exist, and return its
+    node ID."""
+    dsId = _datasetRefId(dsRef)
+    nodeName = allDatasetRefs.get(dsId)
+    if nodeName is None:
+        nodeName = f"DATASET_{len(allDatasetRefs)}"
+        allDatasetRefs[dsId] = nodeName
+        # Simple label: datasetType name and run.
+        label_lines = [f"**{dsRef.datasetType.name}**", f"run: {dsRef.run}"]
+        # Add dataId info.
+        for k in sorted(dsRef.dataId.required.keys()):
+            label_lines.append(f"{k}={dsRef.dataId[k]}")
+        label = "<br>".join(label_lines)
+        print(f'{nodeName}["{label}"]', file=file)
+    return nodeName
+
+
+def graph2mermaid(qgraph: QuantumGraph, file: Any) -> None:
+    """Convert QuantumGraph into a Mermaid flowchart (top-down).
+
+    Parameters
+    ----------
+    qgraph : `lsst.pipe.base.QuantumGraph`
+        QuantumGraph instance.
+    file : `str` or file object
+        File where Mermaid flowchart is written, can be a file name or file
+        object.
+
+    Raises
+    ------
+    OSError
+        Raised if the output file cannot be opened.
+    ImportError
+        Raised if the task class cannot be imported.
+    """
+    # Open a file if needed.
+    close = False
+    if not hasattr(file, "write"):
+        file = open(file, "w")
+        close = True
+
+    # Start Mermaid code block and flowchart.
+    print("```mermaid", file=file)
```

Comment:
This is not part of mermaid, right? This is a markdown wrapper around mermaid, to allow mermaid to be parsed in a markdown file? This may break things for other users who are using another tool to convert mermaid into a PNG/SVG/PDF/etc, and so should probably be removed.

Reply:
Correct, we don't really need that. Removed it!